### ISSUE TITLE
Add E2E test to verify workload cluster upgrades

### ISSUE THREAD

**Issue by srm09** [kubernetes-sigs/cluster-api#4043]
**User Story**

As a developer, I would like to run an E2E test to verify upgrading CAPI workload cluster between 2 supported versions.  

**Detailed Description**

- A new prow job should be run periodically which runs the entire upgrade matrix for workload clusters.
- As for the actual test, we carry out the following steps
    - Create a workload cluster with version, say v.1.17
    - Upgrade the workload cluster to version, say v1.18
    - Run conformance tests on the upgraded cluster to verify the normal working of the cluster post upgrade. 
    - As @fabriziopandini commented below, upgrade refers to upgrading the CP, MachinePool and MachineDeployment worker nodes.

**Anything else you would like to add:**

For the complete discussion on the introduction of these tests, look at https://github.com/kubernetes-sigs/cluster-api/issues/3693#issuecomment-730375355 . 
**TL;DR** We need to add E2E tests to verify the upgrades between supported k8s versions.

/kind feature

**Comment by srm09** [comment]
/assign
Waiting for the PRs #3916 and #4042 to get merged.
cc: @fabriziopandini

**Comment by vincepri** [comment]
Any reason we've picked these two specific Kubenretes versions?

**Comment by detiber** [comment]
> Any reason we've picked these two specific Kubernetes versions?

+1, While I think it makes sense to have *something* as the defaults, I would like to see something that is more easily configured at runtime that would allow for automating prow jobs more easily.

**Comment by srm09** [comment]
The idea was to start with a single job which tests one of the upgrade paths mentioned in the https://github.com/kubernetes-sigs/cluster-api/issues/3693#issuecomment-730375355. Once that gets done, we raise new issues for all the other upgrade tests required and ask some help from the community to finish those off.

>  I would like to see something that is more easily configured at runtime that would allow for automating prow jobs more easily.
That is the end goal, to keep the test/job configurable enough to introduce more upgrade paths easily.

**Comment by fabriziopandini** [comment]
As a reminder, upgrade in this context should include upgrading the control plane and the workers (including machines controlled by MachineDeployments and machines controlled by MachinePools)

**Comment by fabriziopandini** [comment]
#3916 merged, so now this activity is unblocked
/help

**Comment by srm09** [comment]
Updating this story to include the entire test matrix for workload cluster upgrade testing.

**Comment by srm09** [comment]
/retitle Add E2E test to verify workload cluster upgrades

**Comment by vincepri** [comment]
Is this being worked on?

**Comment by fabriziopandini** [comment]
yes
There are PR both in CAPI and test infra ready to merge (I pinged recently in soak you and @CecileRobertMichon)
